src/stan-users-guide/odes.Rmd

```diff
@@ -383,62 +271,164 @@ signature.  The only difference is in the call to the integrator for
+`relative_tolerance`, `absolute_tolerance`, and `max_num_steps` parameters.
+These are the same as the regular function names but with `_tol` appended to
+the end. All three control arguments must be supplied with this signature (there
+are no defaults).
 
 ```
-y_hat = integrate_ode_bdf(sho, y0, t0, ts, theta, x_r, x_i,
-                          rel_tol, abs_tol, max_steps);
+vector[2] y_sim[T] = ode_bdf_tol(sho, y0, t0, ts,
+                                 relative_tolerance,
+                                 absolute_tolerance,
+                                 max_num_steps,
+                                 theta);
 ```
 
-The three control arguments are relative tolerance, absolute
-tolerance, and maximum number of steps.   The default values for
-relative and absolute tolerance are both `1e-6` ($10^{-6}$), and
-the default maximum number of steps is `1e6` ($10^6$).
-
-### Data only for control parameters {-}
+`relative_tolerance` and `absolute_tolerance` control the stepsize by specifying
+a target local error that the solver tries to match, and
+`max_num_steps` specifies the maximum number of steps the solver will
+take between output time points before throwing an error.
 
-The control parameters must be data variables---they can not be
+The control parameters must be data variables -- they cannot be
 parameters or expressions that depend on parameters, including local
 variables in any block other than transformed data and generated
-quantities.  User-defined function arguments may be qualified as only
+quantities. User-defined function arguments may be qualified as only
 allowing data arguments using the `data` qualifier.
 
+For the RK45 solver, the default values for relative and absolute tolerance are
+both $10^{-6}$ and the maximum number of steps between outputs is
+one million. For the BDF and Adams solvers, the relative and absolute
+tolerances are $10^{-10}$ and the maximum number of steps between outputs is
+one hundred million.
+
+### Discontinuous ODE System Function {-}
+
+If there are discontinuities in the ODE system function, it is best
+to integrate the ODE between the discontinuities, stop the solver at each
+one, and restart it on the other side.
+
+Nonetheless, the ODE integrator will attempt to integrate over discontinuities
+it encounters in the state function. The accuracy of the solution near the
+discontinuity may be problematic (requiring many small steps). An example of
+such a discontinuity is a lag in a pharmacokinetic model, where a
+concentration is zero for times $0 < t < t'$ and then positive for $t \geq t'$.
+In this example example, we would use code in the system such as
+
+```
+if (t < t_lag)
+  return 0;
+else
+  ... return non-zero value...;
+```
+
+In general it is better to integrate up to `t_lag` in one solve and
+then integrate from `t_lag` onwards in another. Mathematically, the
+discontinuity can make the problem ill-defined and the numerical integrator
+may behave erratically around it.
```

Comment:
> then it's strongly advisable to output some function values in the vicinity of the discontinuity

I don't think this will improve the behavior of the integrator around those points though. Intermediate outputs are just interpolated from the actual path the solver took.

I guess I could suggest this if there was something to do with the output. Like make plots or something. I just don't think there's an obvious way to diagnose if there are problems or not here.

Reply:
No, it really does make a difference. I had experienced this myself when coding dosing events as approximate delta peaks. Only when I added observation time points around the dosing time, only then the integrator picked up the dosing event. The big disadvantage of the approach you propose to restart is that you must restart the integration process with the initials being parameters whereas if you keep integrating in a single go, then - for pharmacokinetic problems at least - you can use data as the initial. This will dramatically speedup the solve process as the sensitivity system is a lot smaller as compared to making initials varying.